rfc/core language/4956-cluster-pools.md

```diff
@@ -0,0 +1,100 @@
+
+Proposed configuration:
+```
+    clusters:
+       poolNameClusterMap:
+        pool_1:
+        - id: flyte-1-k8s
+          weight: 1
+        - id: flyte-2-k8s
+	      weight: 2
+        pool_2:
+        - id: flyte-3-k8s
+	      weight: 6
+        - id : flyte-4-k8s
+	      weight: 99
+      labelClusterMap:
+        production:
+        - id: flyte-1-k8s
+          weight: 1
+      clusterConfigs:
+      - name: flyte-1-k8s
+        endpoint: 127.0.0.1
+        enabled: true
+        auth:
+          type: file_path
+          tokenPath: /etc/…
+          certPath: /etc/…
+    - name: flyte-2-k8s
+        endpoint: 127.0.0.2
+        enabled: true
+...
+```
+
+At the start of FlyteAdmin it will create a new RandomClusterSelector with a `poolWeightedRandomMap`. 
+
+
+### When an execution is launched
+
+Flyteadmin runs the [Execute](https://github.com/flyteorg/flyte/blob/master/flyteadmin/pkg/workflowengine/impl/k8s_executor.go#L36) method and builds the workflow. Now we need to know which Kubernetes Cluster to hit. We will have the pool information available in the ExecutionData that is already passed into the Execute method. 
+
+**If no clusterPool is defined**: No change in current behavior.
+
+**If clusterPool is defined**: Then we will add the poolName into the [ExecutionTargetSpec](https://github.com/flyteorg/flyte/blob/master/flyteadmin/pkg/executioncluster/execution_target.go#L13-L20) (we will need to modify the struct). This is passed to the getTarget where we check if it has a poolName defined and get the value from the poolWeightedRandomMap. There are still open questions regarding the priority of labels / pool names.
+
+
+## 4 Metrics & Dashboards
+
+If implemented correctly, we should be able to capture where executions are being triggered. This would be extra useful for our use case to make sure we’re routing executions to the correct pool of clusters
+
+## 5 Drawbacks
+
+Other than the usual drawbacks of implementing a new feature (the introduction of new bugs/problems) we don’t see any potential drawbacks with this implementation. The fields are already exposed but they are only missing the implementation. 
+
+We don’t expect any additional overhead for execution or impact in terms of API for flyteadmin.
+
+## 6 Alternatives
+
+As mentioned above, the only alternative we have implemented today is using the [Execution Cluster Labels](https://docs.flyte.org/en/latest/deployment/deployment/multicluster.html#configure-execution-cluster-labels). This can bring further problems when we want to route executions to different clusters from the same Project/Domain.
+
+## 7 Potential Impact and Dependencies
+
+We do not see any.
+
+
+## 8 Unresolved questions
+
+What takes priority Labels or Pools? 
```

Comment:
I would vote for labels to take priority as those seem to be more execution specific. I might be missing something though

Reply:
I think since pools are specified dynamically at execution time I think they should override labels as that is set more "statically"?

Reply:
Yes the label usage seems more related to  "I need this X type of cluster for this <PROJECT/Domain>".

So to be sure, we first check if there's a label set, then under the clusters in that label if there's a pool. If not then default using the weight. (Similar to what is done today.)